### PR TITLE
Refactoring of reverse tunnel agent

### LINKED ISSUE #358: timeouts problem
Sometimes I get this error: 

ERROR: no active connections
ERROR: ssh: subsystem request failed

Let's increase TTLs by a lot

### PR DESCRIPTION
**Pull request by kontsevoy**
- reduced number of goroutines
- reduced number of 'sleep constants', settling on just one:
  `defaults.HeartbeatPeriod`
- increased the interval

Fixes #358

### CODE REVIEW

File: lib/reversetunnel/agent.go [hunk 1]
```diff
@@ -104,45 +104,10 @@ func (a *Agent) Close() error {
...
-				i++
-				continue
-			}
-			return nil
-		}
-	}
+	conn, err := a.connect()
+	// start heartbeat even if error happend, it will reconnect
+	go a.runHeartbeat(conn)
+	return err
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: trace.Wrap

File: lib/reversetunnel/agent.go [hunk 2]
```diff
@@ -328,75 +235,76 @@ func (a *Agent) proxyTransport(ch ssh.Channel, reqC <-chan *ssh.Request) {
...
-
-	go func() {
-		_, err := hb.SendRequest("ping", false, nil)
+// runHeartbeat is a blocking function which runs in a loop sending heartbeats
+// to the given SSH connection.
+//
+func (a *Agent) runHeartbeat(conn *ssh.Client) {
+	heartbeatLoop := func() error {
+		if conn == nil {
+			return trace.Errorf("heartbeat cannot ping: need to reconnect")
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: `trace.BadParameter`

File: lib/reversetunnel/agent.go [hunk 3]
```diff
@@ -328,75 +235,76 @@ func (a *Agent) proxyTransport(ch ssh.Channel, reqC <-chan *ssh.Request) {
...
+		}
+		defer conn.Close()
+		hb, reqC, err := conn.OpenChannel(chanHeartbeat, nil)
 		if err != nil {
-			a.log.Errorf("failed to send heartbeat: %v", err)
-			errC <- err
-			return
+			return err
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: trace.Wrap

File: lib/reversetunnel/srv.go [hunk 4]
```diff
@@ -595,8 +595,8 @@ func (s *tunnelSite) handleHeartbeat(conn *remoteConn, ch ssh.Channel, reqC <-ch
 					return
 				}
 				s.setLastActive(time.Now())
-			case <-time.After(2 * defaults.ReverseTunnelAgentHeartbeatPeriod):
-				conn.markInvalid(trace.ConnectionProblem(nil, "agent missed 2 heartbeats"))
+			case <-time.After(3 * defaults.ReverseTunnelAgentHeartbeatPeriod):
```
Comment (klizhentas) [inline lib/reversetunnel/srv.go]: this should be a constant too

File: lib/reversetunnel/agent.go [hunk 5]
```diff
@@ -180,79 +145,21 @@ func (a *Agent) checkHostSignature(hostport string, remote net.Addr, key ssh.Pub
...
-			}
-		}
-		a.log.Infof("got transport request: %v", nch.ChannelType())
-		ch, req, err := nch.Accept()
-		if err != nil {
-			a.log.Errorf("failed to accept request: %v", err)
-		}
-		go a.proxyTransport(ch, req)
-	}
+	return conn, err
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: trace.Wrap

File: lib/reversetunnel/agent.go [hunk 6]
```diff
@@ -328,75 +235,76 @@ func (a *Agent) proxyTransport(ch ssh.Channel, reqC <-chan *ssh.Request) {
...
-	ticker := time.NewTicker(defaults.ReverseTunnelAgentHeartbeatPeriod)
-	defer ticker.Stop()
-
-	go func() {
-		_, err := hb.SendRequest("ping", false, nil)
+// runHeartbeat is a blocking function which runs in a loop sending heartbeats
+// to the given SSH connection.
+//
+func (a *Agent) runHeartbeat(conn *ssh.Client) {
+	heartbeatLoop := func() error {
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: this should be a separate function

File: lib/reversetunnel/agent.go [hunk 7]
```diff
@@ -328,75 +235,76 @@ func (a *Agent) proxyTransport(ch ssh.Channel, reqC <-chan *ssh.Request) {
...
-				return
-			case <-closeC:
-				log.Infof("asked to exit")
-				return
+			// ssh channel closed:
 			case req := <-reqC:
 				if req == nil {
-					errC <- trace.Errorf("heartbeat: connection closed")
-					return
+					return trace.Errorf("heartbeat: connection closed")
```
Comment (klizhentas) [inline lib/reversetunnel/agent.go]: `trace.ConnectionProblem`

File: lib/reversetunnel/srv.go (hunk 4 above)
Comment (kontsevoy) [inline lib/reversetunnel/srv.go]: can't. tests set it to 50ms to run faster.

File: lib/reversetunnel/agent.go (hunk 7 above)
Comment (kontsevoy) [inline lib/reversetunnel/agent.go]: that function needs an error object + message. I only have message here.

File: lib/reversetunnel/agent.go (hunk 5 above)
Comment (kontsevoy) [inline lib/reversetunnel/agent.go]: these errors aren't returned, they're logged (with line number) and discarded.

**Comment by kontsevoy** [comment]
I've added a couple of missed `trace.Wrap()` calls. Other suggestions make the logic hard to follow. Right now `runHeartbeat()` incapsulates 100% of Agent lifecycle bookkeeping in a neat local context, so you can see without jumping around what can happen to it.
